### PR TITLE
RA Blog Part 1

### PR DESCRIPTION
**Pull request by mruffin**

### CODE REVIEW

File: _posts/2023-17-10-remote-attestation-blog-part1.markdown [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+--- -->
+
+# <h1 style="text-align: center; color:#6B98BF;"> 🚧 A Hitchhiker's Guide to Remote Attestation 🚧</h1>
+<!-- # $${\color{#6B98BF}🚧 A Hitchhikers Guide to Remote Attestation 🚧}$$ -->
+By: Margie Ruffin <br>
+October 17, 2023
+
+Let’s say you have servers somewhere far away, potentially hosted by your company or some other provider, but you sometimes wonder if you could ever fully trust that no one is doing anything sketchy on them. You may ask yourself, how can I prove that my machines have not been tampered with, or how can I prove that they are being tampered with? You may wonder how you can trust them. With Remote Attestation, you don’t have to wonder; you can use existing hardware solutions to prove that the machines can be trusted and are indeed in a secure state. 
+
+This article is Part 1 of this blog series revolving around Remote Attestation and Keylime. In this piece, we will briefly examine the concept of Remote Attestation and several of its components. In the end, I will introduce you to a software solution called Keylime, which you can use to implement this type of attestation.
```
Comment (THS-on) [inline _posts/2023-17-10-remote-attestation-blog-part1.markdown:15]:
```suggestion
This article is Part 1 of this blog series revolving around Remote Attestation and Keylime. In this piece, we will briefly examine the concept of Remote Attestation and several of its components. In the end, I will introduce you to Keylime, which is a software solution you can use to implement this type of attestation.
```

File: _posts/2023-17-10-remote-attestation-blog-part1.markdown [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+This framework allows changes to the user’s computer, the attested (your remote servers), to be detected by authorized parties, the attestors (you on your local servers). In this framework, the attestor machine is already trusted by the customers, while the nodes they are provisioning through the cloud service providers are not. 
+
+Remote attestation can provide different services, such as measured boot attestation and runtime integrity monitoring, using a hardware-based cryptographic root of trust, otherwise known as a Trusted Platform Module (TPM). A TPM is a chip that conforms to the standards of a secure cryptoprocessor, which is a dedicated microcontroller designed to secure hardware through integrated cryptographic keys. It is used to store artifacts that work to authenticate the platform it is on.
```
Comment (THS-on) [inline _posts/2023-17-10-remote-attestation-blog-part1.markdown]:
```suggestion
Remote attestation can provide different services, such as measured boot attestation and runtime integrity monitoring, using a hardware-based cryptographic root of trust, for example as a Trusted Platform Module (TPM). A TPM is a chip that conforms to the standards of a secure cryptoprocessor, which is a dedicated microcontroller designed to secure hardware through integrated cryptographic keys. It is used to store artifacts that work to authenticate the platform it is on. 
```

Note that the TPM can be also implemented on the CPU. TPM is not the only hardware root of trust out there, but the one that we focus on with Keylime

File: _posts/2023-17-10-remote-attestation-blog-part1.markdown [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    <li> What is Secure Boot? </li>
+    <li> What is Measured Boot? </li>
+    <li> Measured Boot and Secure Boot go Hand in Hand </li>
+    <li> Integrity Measurement Architecture (IMA) for Continuous Attestation</li>
+    <li> Introduction to Keylime, A Remote Attestation Software </li>
+</ul>
+
+<h3 style="text-align left; color:#6B98BF;"> What is Secure Boot? </h3>
+
+Trusted boot technologies rely on a “Root of Trust” (RoT), a source that can always be trusted within a cryptographic system. The RoT is usually a piece of code or hardware that’s been hardened well enough that it is unlikely to be compromised. The RoT cannot be modified at all or cannot be modified without cryptographic credentials.
```
Comment (THS-on) [inline _posts/2023-17-10-remote-attestation-blog-part1.markdown]: Root of trust is already mentioned a few lines above. Maybe introduce the abbreviation already there are, keep the the capitalization consistent.

Comment (THS-on) [inline _posts/2023-17-10-remote-attestation-blog-part1.markdown]: If you want you can follow the definition by the TCG:
```
A component that performs one or more security-specific functions, such as
measurement, storage, reporting, verification, and/or update. It is trusted al-
ways to behave in the expected manner, because its misbehavior cannot be
detected (such as by measurement) under normal operation.
```

Source: https://trustedcomputinggroup.org/wp-content/uploads/TCG-Glossary-V1.1-Rev-1.0.pdf

File: _posts/2023-17-10-remote-attestation-blog-part1.markdown [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    <li> Measured Boot and Secure Boot go Hand in Hand </li>
+    <li> Integrity Measurement Architecture (IMA) for Continuous Attestation</li>
+    <li> Introduction to Keylime, A Remote Attestation Software </li>
+</ul>
+
+<h3 style="text-align left; color:#6B98BF;"> What is Secure Boot? </h3>
+
+Trusted boot technologies rely on a “Root of Trust” (RoT), a source that can always be trusted within a cryptographic system. The RoT is usually a piece of code or hardware that’s been hardened well enough that it is unlikely to be compromised. The RoT cannot be modified at all or cannot be modified without cryptographic credentials. 
+
+Unified Extended Firmware Interface (UEFI) Secure Boot is a security measure developed to ensure that a device is booted using only software trusted by the Original Equipment Manufacturer (OEM). The goal is to prevent malicious software from being loaded and executed early in the boot process. During secure boot, the kernel checks if a cryptographic signature, a hash value, on the boot loader matches a pre-loaded value signed by the manufacturer, stored inside the BIOS, that anyone can check using a public key made available.
```
Comment (THS-on) [inline _posts/2023-17-10-remote-attestation-blog-part1.markdown]: The kernel does not have directly something to do with Secure Boot because it is the last component in that chain. (In some cases the UEFI Secure Boot certificates are loaded onto the kernel keyring, but that is a more complex topic). The UEFI firmware does the verification in the first step, then the bootloaders can either use the firmware to check the next component or as in the case with most x86 UEFI Linux systems that use a shim (https://github.com/rhboot/shim) via the shim lock protocol (e.g. https://www.gnu.org/software/grub/manual/grub/html_node/UEFI-secure-boot-and-shim.html).  For simplification we can just say that in Secure Boot that the next component is verified by the component before.

On a normal UEFI implementation you have three major keys/certificates involved:
* Platform Key (PK): Is created by the OEM, allows the enrollment of the KEK
* Key Exchange Key (KEK): Is on most systems one from Microsoft. Allows to update i.e. the DB entries
* Allowed signatures database (DB): certificates actually used to verify the boot components. On x86 UEFI systems you likely find here the Windows CA and the third-party Windows CA.

Details can be found for example here: https://learn.microsoft.com/en-us/windows-hardware/manufacture/desktop/windows-secure-boot-key-creation-and-management-guidance?view=windows-11

Most users only will not interact with those keys, but in many cases you can also enroll your own setup. I don't know what is most common by the cloud providers.

File: _posts/2023-17-10-remote-attestation-blog-part1.markdown [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+
+Unified Extended Firmware Interface (UEFI) Secure Boot is a security measure developed to ensure that a device is booted using only software trusted by the Original Equipment Manufacturer (OEM). The goal is to prevent malicious software from being loaded and executed early in the boot process. During secure boot, the kernel checks if a cryptographic signature, a hash value, on the boot loader matches a pre-loaded value signed by the manufacturer, stored inside the BIOS, that anyone can check using a public key made available. 
+
+Secure boot prevents boot if the two signatures, the preloaded value, and the calculated value during the boot process do not match. If the decrypted hash does not match the new one just taken, the secure boot will fail for reasons such as someone changing the kernel. This process of checking the kernel's signatures occurs before the OS ever runs.
```
Comment (THS-on) [inline _posts/2023-17-10-remote-attestation-blog-part1.markdown]:
```suggestion
Secure Boot prevents boot if the signature cannot be validated against a certificate enrolled in the bootchain. For example if someone changes the kernel to a custom one, it is not signed by the OS vendor. This causes the boot to stop because the bootloader couldn't verify the signature. The process of checking the kernel's signatures occurs before the OS ever runs.
```

File: _posts/2023-17-10-remote-attestation-blog-part1.markdown [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+
+<h3 style="text-align left; color:#6B98BF;"> What is Measured Boot? </h3>
+
+Measured Boot uses the same RoT as Secure Boot; however, unlike Secure Boot, it will check each startup component, including firmware, all the way up to the boot drivers. A hash is taken at each boot process step for the next object(s) in the chain. The hashes are stored inside the TPM so that they can be securely retrieved later to find out what objects were encountered. With Measured Boot, the boot process is never stopped, but it provides the necessary information to detect attacks.
```
Comment (THS-on) [inline _posts/2023-17-10-remote-attestation-blog-part1.markdown]:
```
Measured Boot uses the same RoT as Secure Boot
```

This statement is false. The RoT for Secure Boot is the DB in the firmware. The RoT for Measured Boot is the TPM. 

```suggestion
Unlike Secure Boot, Measured Boot will measure each startup component, including firmware, all the way up to the boot drivers. A hash is taken at each boot process step for the next object(s) in the chain. The hashes are extended to the TPM  and added to a event log so that they can be securely retrieved later to find out what objects were encountered. With Measured Boot, the boot process is never stopped, but it provides the necessary information to detect attacks. 
```

File: _posts/2023-17-10-remote-attestation-blog-part1.markdown [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+
+Secure Boot depends on Measured Boot to guard the UEFI Bios. Measured Boot will check the competent and take a hash of it to be stored for later to help prove whether or not it has been tampered with. The UEFI Bios will prevent an unsigned kernel from booting properly if the hashes taken do not match at the time and will boot the kernel in “lockdown mode,” which will prevent alterations to it. This makes making retroactive alteration of the measured boot output harder. 
+
+Measured Boot guards the UEFI Bios during the secure boot process by taking a hash at its step in the boot process. Secure boot guards the post-boot integrity of the kernel by keeping it in “lockdown mode” in case something suspicious happens. Both are needed to get to a point where we have a completely booted system with a kernel that can be trusted.
```
Comment (THS-on) [inline _posts/2023-17-10-remote-attestation-blog-part1.markdown]: As already discussed the main difference is:

* Secure Boot: verification stage by stage (normally using signatures), boot fails if something cannot be validated
* Measured Boot: each component is measured before executed the boot process generally not stops.

There also exists the terms of verified and trusted boot. The issue seems to be that there is no clear definitions of those terms .

For Remote Attestation Secure Boot is not strictly necessary, because we can just check if the correct components were booted (it can just make it easier for us to make conclusions about that system). For Local Attestation (e.g. BitLocker) we might want to seal secrets against the Secure Boot state, then those technologies are combined.  We do not need the kernel to be trustworthy for Measured Boot because the RoT is the TPM.

File: _posts/2023-17-10-remote-attestation-blog-part1.markdown [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+
+<h3 style="text-align left; color:#6B98BF;"> Integrity Measurement Architecture (IMA) for Continuous Attestation </h3>
+
+The two previously described aspects of Remote Attestation are helpful for a one-time check to see if the server you have provisioned was altered before it was booted and running. But what can you do if you want to continuously make sure that things aren’t being altered in real-time? You can use Linux kernel's Integrity Measurement Architecture (IMA) for that. Implementing IMA with your Remote Attestation framework lets you detect if files have been accidentally or maliciously altered. You can, both remotely and locally, appraise a file's current measurement against a "good" value stored as an extended attribute and enforce local file integrity. IMA collects file hashes and places them in kernel memory where other applications cannot access or modify them. With Secure and Measured Boot enabled, we established a chain of trust from the TPM all the way up to the running kernel. Because the kernel is now trusted, we can trust it to measure files with IMA, thus extending that trust to the files.
```
Comment (THS-on) [inline _posts/2023-17-10-remote-attestation-blog-part1.markdown]: Note here that for IMA the RoT is still the TPM because we extend those measurements into the TPM.

File: _posts/2023-17-10-remote-attestation-blog-part1.markdown [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+
+Secure boot prevents boot if the two signatures, the preloaded value, and the calculated value during the boot process do not match. If the decrypted hash does not match the new one just taken, the secure boot will fail for reasons such as someone changing the kernel. This process of checking the kernel's signatures occurs before the OS ever runs.
+
+<h3 style="text-align left; color:#6B98BF;"> What is Measured Boot? </h3>
+
+Measured Boot uses the same RoT as Secure Boot; however, unlike Secure Boot, it will check each startup component, including firmware, all the way up to the boot drivers. A hash is taken at each boot process step for the next object(s) in the chain. The hashes are stored inside the TPM so that they can be securely retrieved later to find out what objects were encountered. With Measured Boot, the boot process is never stopped, but it provides the necessary information to detect attacks. 
+
```
Comment (THS-on) [inline _posts/2023-17-10-remote-attestation-blog-part1.markdown:54]: Might be a good idea on how to describe how the PCR extend operation in a TPM works and why it allows us to verify an entire log.

File: _posts/2023-17-10-remote-attestation-blog-part1.markdown [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+In just a few short steps, we can see how IMA works.
+
+<ol>
+    <li> An integrity challenge mechanism allows for a remote attestor (you on your trusted machine) to request a measurement list (IMA Log) and a TPM-signed aggregate of the measurement from the attesting system (the untrusted machine). </li>
```
Comment (THS-on) [inline _posts/2023-17-10-remote-attestation-blog-part1.markdown]: Use here what a PCR is and that we can get those values and and a signature of those values aggregated into one hash.

File: _posts/2023-17-10-remote-attestation-blog-part1.markdown [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+
+Secure Boot depends on Measured Boot to guard the UEFI Bios. Measured Boot will check the competent and take a hash of it to be stored for later to help prove whether or not it has been tampered with. The UEFI Bios will prevent an unsigned kernel from booting properly if the hashes taken do not match at the time and will boot the kernel in “lockdown mode,” which will prevent alterations to it. This makes making retroactive alteration of the measured boot output harder. 
+
+Measured Boot guards the UEFI Bios during the secure boot process by taking a hash at its step in the boot process. Secure boot guards the post-boot integrity of the kernel by keeping it in “lockdown mode” in case something suspicious happens. Both are needed to get to a point where we have a completely booted system with a kernel that can be trusted. 
+
```
Comment (THS-on) [inline _posts/2023-17-10-remote-attestation-blog-part1.markdown:62]: Here it feels like that there is a section missing on how Remote Attestation + Measured Boot work.

File: _posts/2023-17-10-remote-attestation-blog-part1.markdown (hunk 8 above)
Comment (THS-on) [inline _posts/2023-17-10-remote-attestation-blog-part1.markdown]: In IMA appraisal normally means local (in the sense of on the device) file integrity enforcement. Maybe only introduce only the measurement part first, because the appraisal is explained in the next section in more detail.

File: _posts/2023-17-10-remote-attestation-blog-part1.markdown (hunk 3 above)
Comment (THS-on) [inline _posts/2023-17-10-remote-attestation-blog-part1.markdown]: Note that you are mixing trusted and measured boot in the article. Maybe just use measured boot.

File: _posts/2023-17-10-remote-attestation-blog-part1.markdown [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+Remote attestation is a way for you to create peace of mind that you can and will detect when something is amiss inside a cluster. By building a chain of trust from hardware to software, you can ensure your data is protected. 
+
+Keylime is commonly used either on bare metal hardware or in VMs where the TPM is emulated but, from the VM side, treated the same as a hardware TPM. In this article, we have seen what Keylime can detect:
+
+<ul>
+    <li> Modification of the boot process </li>
+    <li> Runtime file and system integrity </li>
+</ul>
+
+While it won’t stop anything after an alert is raised, Keylime is still a great tool to have because it will tell you what has happened and where. Perfect for someone managing a lot of machines!
```
Comment (THS-on) [inline _posts/2023-17-10-remote-attestation-blog-part1.markdown]:
```
While it won’t stop anything after an alert is raised,
```

You can implement what ever you want around the revocation hook that Keylime provides, e.g. remove the node from the cluster when it Keylime detects that it is compromised.

File: _posts/2023-17-10-remote-attestation-blog-part1.markdown [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+
+To use Keylime’s Runtime Integrity Monitoring, there are a few steps that you will need to take to either enable Linux IMA or ensure that it is already running. As we learned a few sections ago, IMA is used to measure the contents of files at runtime. However, before it can do that, you will have to tell it what to measure using an ***ima-policy***. In the Keylime repository, you will find an example policy that can be used, and in the documentation, you will find instructions on enabling IMA. 
+
+In Keylime, a ***runtime policy*** is a list of “golden” cryptographic hashes of files in their untampered state that are used for IMA verification. The Keylime verifier will load the runtime policy and use it to validate the agent’s file states that are loaded onto the TPM. If the files have been tampered with, the hashes will not match, and Keylime will place the agent into a failed state. The Keylime agent and operating system itself are not deemed trustworthy by default. Given the setup only after the successful initial attestation is the system deemed trustworthy, but it still can leave the trusted state at any moment and is, therefore, continuously attested.
```
Comment (THS-on) [inline _posts/2023-17-10-remote-attestation-blog-part1.markdown]: We also support checking against IMA signatures.

**Review by THS-on — COMMENTED** [review]
@mruffin here the comments. If you have questions either add them here or ping me on Slack.

File: _posts/2023-17-10-remote-attestation-blog-part1.markdown (hunk 10 above)
Comment (mruffin) [reply _posts/2023-17-10-remote-attestation-blog-part1.markdown]: I explained in detail how to extend the PCRs in a section above. Do you think I should reiterate or do you mean explain something else?

File: _posts/2023-17-10-remote-attestation-blog-part1.markdown [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+
+This framework allows changes to the user’s computer, the attested (your remote servers), to be detected by authorized parties, the attestors (you on your local servers). In this framework, the attestor machine is already trusted by the customers, while the nodes they are provisioning through the cloud service providers are not. 
+
+Remote attestation can provide different services, such as measured boot attestation and runtime integrity monitoring, using a hardware-based cryptographic "Root of Trust" (RoT), for example, a Trusted Platform Module (TPM). A TPM is a chip that conforms to the standards of a secure cryptoprocessor, which is a dedicated microcontroller designed to secure hardware through integrated cryptographic keys. It is used to store artifacts that work to authenticate the platform it is on. 
+
+
+<div style="text-align: center;">
+<img  src="/Users/mruffin2/Desktop/Research/Keylime/keylime.github.io/assets/images/RemoteAttestation.png" alt="Remote Attestation Diagram">
```
Comment (THS-on) [inline _posts/2023-17-10-remote-attestation-blog-part1.markdown]: The images need to be relative paths, not absolute ones otherwise they won't show up on the website.

File: _posts/2023-17-10-remote-attestation-blog-part1.markdown [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+    <li> What is Secure Boot? </li>
+    <li> What is Measured Boot? </li>
+    <li> Measured Boot and Secure Boot go Hand in Hand </li>
+    <li> Integrity Measurement Architecture (IMA) for Continuous Attestation</li>
+    <li> Introduction to Keylime, A Remote Attestation Software </li>
+</ul>
+
+<h3 style="text-align left; color:#6B98BF;"> What is Secure Boot? </h3>
+
+Measured boot technologies rely on a RoT, a source that can always be trusted within a cryptographic system. The RoT is a component that performs one or more security-specific functions, such as measurement, storage, reporting, verification, and/or update. It is trusted always to behave in the expected manner, because its misbehavior cannot be detected (such as by measurement) under normal operation. The RoT cannot be modified at all or cannot be modified without cryptographic credentials.
```
Comment (THS-on) [inline _posts/2023-17-10-remote-attestation-blog-part1.markdown]: This is in the wrong section.

File: _posts/2023-17-10-remote-attestation-blog-part1.markdown (hunk 10 above)
Comment (THS-on) [reply _posts/2023-17-10-remote-attestation-blog-part1.markdown]: No that's fine

File: _posts/2023-17-10-remote-attestation-blog-part1.markdown [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+
+<h3 style="text-align left; color:#6B98BF;"> Measured Boot and Secure Boot Go Hand in Hand </h3>
+
+These two processes go hand-in-hand to ensure a trusted Operating System (O/S) boots. Measured Boot guards the system from the processor powering on to the point where the operating system is ready to run. The issue is that once the O/S is booted, Measured Boot stops and its output is encapsulated into the measured boot log and the PCR values in the TPM. Now that the O/S is unguarded, it is **conceivable** that someone could manage to navigate to the boot event log, the PCRs, and the TPM and alter them retroactively. They could even go as far as to replace the TPM device with a virtual TPM.
```
Comment (THS-on) [inline _posts/2023-17-10-remote-attestation-blog-part1.markdown]: The word guards has thrown me off here, because it doesn't do it actively, instead it provides a way that we can check any tampering of the system afterwards.

Also add a sentence why it doesn't mater if someone switches out the TPM. Some mention of the Endorsement Key + Certificate as a unique identity.

File: _posts/2023-17-10-remote-attestation-blog-part1.markdown [hunk 17]
```diff
@@ -0,0 +1,139 @@
...
+
+<h3 style="text-align left; color:#6B98BF;"> Measured Boot and Secure Boot Go Hand in Hand </h3>
+
+These two processes go hand-in-hand to ensure a trusted Operating System (O/S) boots. Measured Boot guards the system from the processor powering on to the point where the operating system is ready to run. The issue is that once the O/S is booted, Measured Boot stops and its output is encapsulated into the measured boot log and the PCR values in the TPM. Now that the O/S is unguarded, it is **conceivable** that someone could manage to navigate to the boot event log, the PCRs, and the TPM and alter them retroactively. They could even go as far as to replace the TPM device with a virtual TPM. 
+
+Secure Boot depends on Measured Boot to guard the UEFI Bios. Measured Boot will check the competent and take a hash of it to be stored for later to help prove whether or not it has been tampered with. The UEFI Bios will prevent an unsigned kernel from booting properly if the hashes taken do not match at the time and will boot the kernel in “lockdown mode,” which will prevent alterations to it. This makes making retroactive alteration of the measured boot output harder.
```
Comment (THS-on) [inline _posts/2023-17-10-remote-attestation-blog-part1.markdown]: No only secure boot does this all. It just when Secure Boot does a decision it gets measured by measured boot.

>This makes making retroactive alteration of the measured boot output harder.

Yes, but the TPM is the reason that it doesn't matter, not secure boot.

Secure Boot -> Local integrity checking at the point of execution
Measured Boot -> Allows checking the integrity after the system was bootet.

File: _posts/2023-17-10-remote-attestation-blog-part1.markdown [hunk 18]
```diff
@@ -0,0 +1,139 @@
...
+
+Measured Boot guards the UEFI Bios during the secure boot process by taking a hash at its step in the boot process. Secure boot guards the post-boot integrity of the kernel by keeping it in “lockdown mode” in case something suspicious happens. Both are needed to get to a point where we have a completely booted system with a kernel that can be trusted. 
+
+In any production enviroment a large number of different types of nodes can be found. One can use remote attestation to in cobination with measured boot to determine and verifiy a platform's configuration. A measured boot reference state can be specified ahead of time for each node type and given to the remote attestation operator along with a measured boot policy that is used to instruct the verifier on how to do the comparison. With this mechanism in place the operator can ensure the validity of the kernals for their enitre cluster.
```
Comment (THS-on) [inline _posts/2023-17-10-remote-attestation-blog-part1.markdown]:
```suggestion
In any production enviroment a large number of different types of nodes can be found. One can use remote attestation to in combination with measured boot to determine and verify a platform's configuration. A measured boot reference state can be specified ahead of time for each node type and given to the remote attestation operator along with a measured boot policy that is used to instruct the verifier on how to do the comparison. With this mechanism in place the operator can ensure the validity of the kernels for their entire cluster. 
```

File: _posts/2023-17-10-remote-attestation-blog-part1.markdown (hunk 17 above)
Comment (mruffin) [reply _posts/2023-17-10-remote-attestation-blog-part1.markdown]: Should I take the last sentence out since it isn't accurate or rephrase it to include the reason being the TPM?

Comment (mruffin) [reply _posts/2023-17-10-remote-attestation-blog-part1.markdown]: I can also add a sentence after the ***which will prevent alterations to it.*** to say that measured boot will allow for integrity checking after the system is booted

**Comment by mpeters** [comment]
@mruffin just a couple of house keeping items before this can be merged: first, you'll need to sign your commit using the `-s` option to `git`. This can be done as part of a `git amend -s ...` if you'd like. This is a requirement for our project since we're part of the CNCF.

And last would be fixing any conflicts since github won't let us rebase and merge with conflicts.

**Comment by stringlytyped** [comment]
I just had a read through and this is an excellent introduction to the topic. Well done @mruffin! I have never had the chance to dig into IMA deeply at all, so I found those bits particularly illuminating. I don't wish to delay acceptance of this PR but I did notice one markdown formatting issue which would be good to fix before publishing. I'll highlight in a separate comment

File: _posts/2023-17-10-remote-attestation-blog-part1.markdown [hunk 19]
```diff
@@ -0,0 +1,138 @@
...
+    <li> Introduction to Keylime, A Remote Attestation Software </li>
+</ul>
+
+<h3 style="text-align left; color:#6B98BF;"> What is Secure Boot? </h3>
+
+Unified Extended Firmware Interface (UEFI) Secure Boot is a security measure developed to ensure that a device is booted using only software trusted by the Original Equipment Manufacturer (OEM). The goal is to prevent malicious software from being loaded and executed early in the boot process. During secure boot, the next component is verified by the component before.
+
+Secure Boot prevents boot if the signature cannot be validated against a certificate enrolled in the bootchain. For example if someone changes the kernel to a custom one, it is not signed by the OS vendor. This causes the boot to stop because the bootloader couldn't verify the signature. The process of checking the kernel's signatures occurs before the OS ever runs.
+
+<h3 style="text-align left; color:#6B98BF;"> What is Measured Boot? </h3>
```
Comment (stringlytyped) [inline _posts/2023-17-10-remote-attestation-blog-part1.markdown:49]: Very minor issue but it seems that because there's not an empty line between the H3 (line 49) and the opening paragraph of the section (line 50), at least on GH, line 50 is not interpreted as a paragraph and does not get wrapped in `<p>` tags. Not sure if the markdown parser used to render the blog posts acts differently though

Comment (mruffin) [reply _posts/2023-17-10-remote-attestation-blog-part1.markdown:49]: I added a space, so hopefully that clears things up.

File: _posts/2023-17-10-remote-attestation-blog-part1.markdown [hunk 20]
```diff
@@ -0,0 +1,144 @@
+<!-- ---
+layout: blog
+title:  "A Hitchhikers Guide to Remote Attestation"
```
Comment (mheese) [inline _posts/2023-17-10-remote-attestation-blog-part1.markdown:3]: love the title Margie :)

File: _posts/2023-17-10-remote-attestation-blog-part1.markdown [hunk 21]
```diff
@@ -0,0 +1,144 @@
...
+
+Remote attestation can provide different services, such as measured boot attestation and runtime integrity monitoring, using a hardware-based cryptographic "Root of Trust" (RoT), for example, a Trusted Platform Module (TPM). A TPM is a chip that conforms to the standards of a secure cryptoprocessor, which is a dedicated microcontroller designed to secure hardware through integrated cryptographic keys. It is used to store artifacts that work to authenticate the platform it is on. 
+Remote attestation can provide different services, such as measured boot attestation and runtime integrity monitoring, using a hardware-based cryptographic "Root of Trust" (RoT), for example, a Trusted Platform Module (TPM). A TPM is a chip that conforms to the standards of a secure cryptoprocessor, which is a dedicated microcontroller designed to secure hardware through integrated cryptographic keys. It is used to store artifacts that work to authenticate the platform it is on.
```
Comment (mheese) [inline _posts/2023-17-10-remote-attestation-blog-part1.markdown]: Maybe worth noting that the recent trend for TPMs goes towards firmware TPMs, and that there are pros and cons to that. Most hardware TPMs are I2C connected which has its own problems/attack vectors.

**Comment by mheese** [comment]
@mruffin are we releasing this blog post soon? or do you need some more deeper reviews?

**Comment by mruffin** [comment]
> @mruffin are we releasing this blog post soon? or do you need some more deeper reviews?

@mheese I think we are just about there. Thore wanted to do one more glance over after his defense before publishing.

File: _posts/2023-17-10-remote-attestation-blog-part1.markdown [hunk 22]
```diff
@@ -0,0 +1,143 @@
...
+
+Secure Boot depends on Measured Boot to guard the UEFI Bios. Measured Boot will check the component and take a hash of it to be stored for later to help prove whether or not it has been tampered with. The UEFI Bios will prevent an unsigned kernel from booting properly if the hashes taken do not match at the time and will boot the kernel in “lockdown mode,” which will prevent alterations to it. Measured boot will allow for integrity checking after the system is booted.
```
Comment (THS-on) [inline _posts/2023-17-10-remote-attestation-blog-part1.markdown]: Just write UEFI instead of UEFI Bios.

These are three mostly orthogonal technologies:
- Secure Boot: Enforce integrity locally using signatures
- Measured Boot: Make the boot chain and the security related settings attestable once the system is booted
- Kernel lockdown features: Enforce higher integrity measures in the kernel. (Require modules to be signed, lock down ACPI features etc.)

Basically Measured Boot allows us to verify the configuration of Secure Boot after the fact. Similarly Secure Boot ensures that in the first place only software that we trust was booted (assuming nobody tampered with the UEFI CA setup)

File: _posts/2023-17-10-remote-attestation-blog-part1.markdown [hunk 23]
```diff
@@ -0,0 +1,143 @@
...
+Using a hardware-based cryptographic RoT, a Keylime user can monitor their remote nodes for system tampering at boot while continuously in service. Keylime is an alert system that will notify the operator by raising a flag if measured boot attestation fails (i.e., someone has altered the kernel) or if runtime integrity monitoring fails (i.e., someone maliciously altered an executable system file). 
+
+The Keylime tool mainly consists of 4 components: an **agent**, a **verifier**, a **registrar**, and a commandline tool called the **tenant**. 
+
+The agent is a service that runs on an untrusted operating system, which is to be attested. It communicates with the TPM and collects necessary data to make attestation possible. 
+
+The verifier is in charge of implementing the attestation of the agent. Essentially, it is the component saying ***yes, the agent is in policy***, or ***no, the agent is not in policy***. If the agent fails attestation, the verifier can also send a revocation message (i.e., raise a flag) to alert the Keylime operator of a failed state.
```
Comment (THS-on) [inline _posts/2023-17-10-remote-attestation-blog-part1.markdown]:
```suggestion
The verifier is in charge of implementing the attestation of the agent. Essentially, it is the component saying ***yes, the agent is trusted***, or ***no, the agent is not trusted***. If the agent fails attestation, the verifier can also send a revocation message (i.e., raise a flag) to alert the Keylime operator of a failed state. 
```

File: _posts/2023-17-10-remote-attestation-blog-part1.markdown [hunk 24]
```diff
@@ -0,0 +1,143 @@
...
+
+In Keylime, a ***runtime policy*** is a list of “golden” cryptographic hashes of files in their untampered state used for IMA verification. The Keylime verifier will load the runtime policy and use it to validate the agent’s file states that are loaded onto the TPM. If the files have been tampered with, the hashes will not match, and Keylime will place the agent into a failed state. The Keylime agent and operating system itself are not deemed trustworthy by default. Given the setup only after the successful initial attestation, the system is deemed trustworthy, but it still can leave the trusted state at any moment and is, therefore, continuously attested. It is important to mention that Keylime also supports the verification of IMA file signatures, which also helps to detect modifications on immutable files and can be used to complement or even replace the allowlist of hashes in the runtime policy if all relevant executables and libraries are signed. However, this setup is beyond the scope of this work.
```
Comment (THS-on) [inline _posts/2023-17-10-remote-attestation-blog-part1.markdown]:
```suggestion
In Keylime, a ***runtime policy*** is a list of “golden” cryptographic hashes of files in their untampered state used for IMA verification. The Keylime verifier will load the runtime policy and use it to validate the agent’s file states that are loaded onto the TPM. If the files have been tampered with, the hashes will not match, and Keylime will place the agent into a failed state. The Keylime agent and operating system itself are not deemed trustworthy by default. Given the setup only after the successful initial attestation, the system is deemed trustworthy, but it still can leave the trusted state at any moment and is, therefore, continuously attested. It is important to mention that Keylime also supports the verification of IMA file signatures, which also helps to detect modifications on immutable files and can be used to complement or even replace the allowlist of hashes in the runtime policy if all relevant executables and libraries are signed. However, this setup is beyond the scope of this introduction.
```

**Review by THS-on — CHANGES_REQUESTED** [review]
The major thing is with how Secure and Measured Boot are tied together. Otherwise this looks really nice and I only added a few suggestions

**Comment by mheese** [comment]
@THS-on why did we close this? Are we not publishing this blog post anymore?

**Comment by mruffin** [comment]
@mheese, it was pushed through another branch using a single commit. Here is the [link to the post.](https://keylime.dev/blog/2024/02/07/remote-attestation-blog-part1.html )
